services/checks/checks.go

```diff
@@ -308,26 +351,28 @@ func (s *Service) minPMMAgentVersion(t check.Type) *version.Parsed {
 func (s *Service) executeChecks(ctx context.Context) {
 	s.l.Info("Executing checks...")
 
-	var alertsIDs []string
+	var checkResults []sttCheckResult
 
-	mySQLAlertsIDs := s.executeMySQLChecks(ctx)
-	alertsIDs = append(alertsIDs, mySQLAlertsIDs...)
+	mySQLCheckResults := s.executeMySQLChecks(ctx)
+	checkResults = append(checkResults, mySQLCheckResults...)
 
-	postgreSQLAlertsIDs := s.executePostgreSQLChecks(ctx)
-	alertsIDs = append(alertsIDs, postgreSQLAlertsIDs...)
+	postgreSQLCheckResults := s.executePostgreSQLChecks(ctx)
+	checkResults = append(checkResults, postgreSQLCheckResults...)
 
-	mongoDBAlertsIDs := s.executeMongoDBChecks(ctx)
-	alertsIDs = append(alertsIDs, mongoDBAlertsIDs...)
+	mongoDBCheckResults := s.executeMongoDBChecks(ctx)
+	checkResults = append(checkResults, mongoDBCheckResults...)
 
-	// removing old STT alerts except created during current run
-	s.alertsRegistry.RemovePrefix(alertsPrefix, sliceToSet(alertsIDs))
+	s.alertsRegistry.clean()
+	for _, result := range checkResults {
+		s.createAlert(result.checkName, result.target, &result.result)
+	}
```

Comment:
Ticket specifically says "Replace checks stored in the registry in executeChecks method **atomically**." StartChecks API can be called concurrently, and we will have a race here.

I think that complexity should be in registry: remove `clear()` method, add `set(alerts ammodels.PostableAlerts)` method that replaces alerts

Reply:
> Ticket specifically says "Replace checks stored in the registry in executeChecks method atomically." StartChecks API can be called concurrently, and we will have a race here.

registry uses a mutex before mutating the alert map won't that be enough to avoid the race?

Reply:
Registry uses mutex for each individual operation. That's not enough. Consider goroutine 1 being on that line, and goroutine 2 calling `clean()` on line 365. There no data race that Go race detector can detect, but the registry is empty and data is lost.

Reply:
But that situation covered by alert TTL that three times longer than resend interval. Technically on UI such situation will be unnoticed, isn't it?

Reply:
Yeah, until we add API to get current check results (not alerts) :)

Reply:
But then "it is a completely different story" )

Sorry @palash25 for that solution

Reply:
Yes, we have that different story. I'm not saying that we should implement that API now. But we also should not make something that will make it harder in the future. The biggest reason for that ticket is to have a registry of check results, not alerts – exactly because we will add APIs later.

I think this part can be reworked something like that:
1. Make `createAlert` a function returning alert object.
2. Replace both `clean` and `createAlert` methods with `set(alerts []portableAlert)` method that sets the whole slice.
3. `collect()` will still return a copy of that slice.